File: python/puzzles/euler11.py
# Largest product in a grid
"""
In the 20×20 grid below, four numbers along a diagonal line have been marked in red.

The product of these numbers is 26 × 63 × 78 × 14 = 1788696.

What is the greatest product of four adjacent numbers in the same direction (up, down, left, right, or diagonally) in the 20×20 grid?

"""

import numpy as np
from python.tools import utils, profile
from python.tools.primes import product


@profile.function
def euler11better():
    grid = utils.load_puzzle_input("euler11")

    grid = [g.split(" ") for g in grid.strip().split("\n")]
    grid = np.array(grid).astype(int)

    span = 4
    max_prod = 0
    for ii in range(grid.shape[0] - span):
        for jj in range(grid.shape[1] - span):
            for vector in [
                grid[ii, jj : jj + span],  # horizontal
                grid[ii : ii + span, jj],  # vertical
                grid[range(ii, ii + span), range(jj, jj + span)],  # diagonal
                grid[range(ii, ii - span, -1), range(jj, jj + span)],  # diagonal up
            ]:

                if (prod := product(vector)) > max_prod:
                    max_prod = prod

    return max_prod


@profile.function
def euler11old():
    grid = utils.load_puzzle_input("euler11")

    grid = [g.split(" ") for g in grid.strip().split("\n")]  # split string into entries
    del grid[0]
    grid = np.array(grid).astype(int)

    span = 4
    products = set()
    # scan horizontal spans
    for ii in range(grid.shape[0]):  # scroll through the rows
        for jj in range(grid.shape[1]):  # scroll through the columns
            if jj + span <= grid.shape[1]:  # horizontal
                vector = grid[ii, jj : jj + span]
                products.add(product(vector))
            if ii + span <= grid.shape[0]:  # vertical
                vector = grid[ii : ii + span, jj]
                products.add(product(vector))
            if (ii + span <= grid.shape[0]) and (jj + span <= grid.shape[1]):  # diag down
                vector = grid[range(ii, ii + span), range(jj, jj + span)]
                products.add(product(vector))
            if (ii - span >= 0) and (jj + span <= grid.shape[1]):  # diag up
                vector = grid[range(ii, ii - span, -1), range(jj, jj + span)]
                products.add(product(vector))

    return max(products)


def euler11():
    return euler11better()


if __name__ == "__main__":
    assert euler11() == 70600674

File: go/puzzles/euler11.go
package puzzles

import (
	"euler/utils"
	"fmt"
	"strconv"
	"strings"
)

func parseGrid() [20][20]int {
	raw := utils.LoadPuzzleInput("euler11")
	rows := strings.Split(raw, "\n")
	const N = 20
	grid := [N][N]int{}
	for y, row := range rows {
		if len(row) == 0 {
			continue
		}
		rowStrings := strings.Fields(row)
		for x, numStr := range rowStrings {
			num, err := strconv.Atoi(string(numStr))
			if err != nil {
				panic(len(row))
			}
			grid[y][x] = num
		}
	}
	return grid
}

func euler11Brute() string {
	const N = 20
	grid := parseGrid()
	span := 4
	maxProd := 0
	for y := 0; y < N-span; y++ {
		for x := 0; x < N-span; x++ {
			hor := [4]int{}
			ver := [4]int{}
			diag := [4]int{}
			diagUp := [4]int{}
			for ii := 0; ii < 4; ii++ {
				hor[ii] = grid[y][x+ii]         // horizontal
				ver[ii] = grid[y+ii][x]         // vertical
				diag[ii] = grid[y+ii][x+ii]     // diagonal down
				diagUp[ii] = grid[y+4-ii][x+ii] // diagonal up
			}
			vectors := [][4]int{hor, ver, diag, diagUp}
			for _, vector := range vectors {
				prod := vector[0] * vector[1] * vector[2] * vector[3]
				if prod > maxProd {
					maxProd = prod
				}
			}
		}
	}
	return fmt.Sprint(maxProd)
}

func Euler11() string {
	return euler11Brute()
}
